merlin/metric.py

```diff
@@ -0,0 +1,134 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+# TODO add serializer and desirealiser from and to json
+
+from pyspark.sql.types import StructType, StructField, StringType, LongType, BooleanType, DoubleType, ArrayType, \
+    MapType
+from merlin.values import ComposedValue
+from merlin.stage import Stage
+
+class Metric:
+    """
+    Basic definition of a metric
+    """
+
+    def __init__(self, id: str, w_time: int, fun_expr: str, h_lvl: int, v_lvl: int):
+        self.id = id
+        self.time = None
+        self.w_time = w_time
+        self.val = None
+        self.group_map = {}
+        self.group_keys = []
+        self.fun_expr = fun_expr
+        self.fun_vars = {}
+        self.h_lvl = h_lvl
+        self.v_lvl = v_lvl
+        self.compute_time = None
+
+
+class OutputMetric(object):
+    __slots__ = ["id", "time", "w_time", "val", "h_lvl", "v_lvl",
+                 "group_map",
+                 "group_keys",
+                 "func_expr",
+                 "func_vars",
+                 "compute_date",
+                 "compute_hour"]
+
+    SPARK_TYPE = MapType(StringType(), StructType([
+        StructField("type", StringType(), True),
+        StructField("int", LongType(), True),
+        StructField("double", DoubleType(), True),
```

Comment:
To be consistent and more specific, can we rename some things?
* change `int` -> `int64`
* change `double` -> `float64`
* add `json` ? (if possible)

For the vectors, I propose to simply add `s`
* `i_vec` -> `int64s`
* `d_vec` -> `float64s`
* `b_vec` -> `bools`
* `s_vec` -> `strings`

For the maps, prefix with `map`
* `i_map` -> `map_int64`
* `d_map` -> `map_float64`
* `b_map` -> `map_bool`
* `s_map` -> `map_string`

Reply:
There is no json  
https://spark.apache.org/docs/1.6.0/api/java/org/apache/spark/sql/types/package-summary.html
but we can set `type = "json"` than `json =  string json encoded`
What do you think?

Reply:
also I am thinking when we write the queries to use the metrics I think it would be easier to 
write `int` rather than `int64`
e.g.

```
select grop_map['sensor_age'].int

```
rather 
```
 select grop_map['sensor_age'].int64
```

Reply:
Regarding the json 
My idea is the following

```
o = ComposedValue({'a': 1, 'b': [1, 2, 3], 'c': False}, "json_test",
                          serialize=True, serialize_type=SerializedTypes.JSON, serializer=json.dumps
```
```

{'type': 'json', 'long': None, 'double': None, 'bool': None, 'string': '{"a": 1, "b": [1, 2, 3], "c": false}', 'vec_long': None, 'vec_double': None, 'vec_bool': None, 'vec_string': None, 'map_long': None, 'map_double': None, 'map_bool': None, 'map_string': None}
```

Reply:
For `int`, maybe we can do just `int` and `float` (instead of `double`) and assume everything is 64-bit precision. Alternatively we can use `i64` and `f64` (esp if we will have `i32` and `f32` types in future?

Reply:
I like the `ComposedValue` idea, but I would use slightly different naming:
```
o = StructValue({'a': 1, 'b': [1, 2, 3], 'c': False}, "json_test", encoding=Codec.JSON, encoder=json.dumps)
```

`struct` is the term used in `orc`, I assume if you specify an `encoding` that `serialize=True` is auto-added. Similarly, I would add a default encoder so that we could write simply:

```
o = StructValue({'a': 1, 'b': [1, 2, 3], 'c': False}, "json_test", encoding=Codec.JSON)
```

WDYT?

Reply:
Yes so now it looks quite clean 

```
        source_metric = SourceMetric(metric_id="volumetric_flow_rate",
                                     time_window=500,
                                     func_expr="mass / density / time",
                                     horizontal_level=1,
                                     vertical_level=0,
                                     version="1.0",
                                     func_vars=["mass", "density", "time"])

        mass = StructuredValue(25.0, "mass")
        density = StructuredValue(5.0, "density")
        location = StructuredValue({"lat": 120.0, "long": 10.0}, "location")
        output_metric = OutputMetric(source_metric, 5.0, datetime(2020, 3, 25, 1, 0))
        output_metric.add_func_var(mass)
        output_metric.add_func_var(density)
        output_metric.add_group_value(location)


```
output metric looks like

```
{
    "id": "volumetric_flow_rate",
    "measure_time": 1585069200,
    "compute_time": 1586784256,
    "compute_date": "2020-04-13",
    "compute_hour": 21,
    "time_window": 500,
    "val": 5.0,
    "group_map": {
        "location": {
            "type": "map_double",
            "long": null,
            "double": null,
            "bool": null,
            "string": null,
            "vec_long": null,
            "vec_double": null,
            "vec_bool": null,
            "vec_string": null,
            "map_long": null,
            "map_double": {
                "lat": 120.0,
                "long": 10.0
            },
            "map_bool": null,
            "map_string": null
        }
    },
    "group_keys": [
        "location"
    ],
    "func_vars": {
        "mass": {
            "type": "double",
            "long": null,
            "double": 25.0,
            "bool": null,
            "string": null,
            "vec_long": null,
            "vec_double": null,
            "vec_bool": null,
            "vec_string": null,
            "map_long": null,
            "map_double": null,
            "map_bool": null,
            "map_string": null
        },
        "density": {
            "type": "double",
            "long": null,
            "double": 5.0,
            "bool": null,
            "string": null,
            "vec_long": null,
            "vec_double": null,
            "vec_bool": null,
            "vec_string": null,
            "map_long": null,
            "map_double": null,
            "map_bool": null,
            "map_string": null
        }
    },
    "func_expr": "mass / density / time",
    "horizontal_level": 1,
    "vertical_level": 0,
    "version": "1.0"
}
                                                         [
```

Reply:
Much better indeed